alphacsc/_encoder.py

```diff
@@ -0,0 +1,336 @@
+        'D_hat should be a valid array of shape ' \
+        '(n_trials, n_channels, n_times) ' \
+        'or (n_atoms, n_channels + atom_support).'
+
+    assert algorithm in ['batch', 'greedy', 'online', 'stochastic'], \
+        f'unrecognized algorithm type: {algorithm}.'
+
+    # ASK can reg be less than 0?
+    assert reg is not None, 'reg value cannot be None.'
+
+    assert loss in ['l2', 'dtw', 'whitening'], \
+        f'unrecognized loss type: {loss}.'
+
+    assert isinstance(
+        loss_params, dict), 'loss_params should be a valid dictionary.'
+
+    assert uv_constraint in ['joint', 'separate'], \
+        f'unrecognized uv_constraint type: {uv_constraint}.'
+    # ASK check invalid values for n_atoms, atom_support, n_jobs
+
+    if solver in ['l-bfgs', 'lgcd']:
+        return AlphaCSCEncoder(
+            solver,
+            z_kwargs,
+            X,
+            D_hat,
+            n_atoms,
+            atom_support,
+            algorithm,
+            reg,
+            loss,
+            loss_params,
+            uv_constraint,
+            feasible_evaluation,
+            n_jobs,
+            use_sparse_z)
+    else:
+        raise ValueError(f'unrecognized solver type: {solver}.')
+
+
+class BaseZEncoder:
+    def compute_z(self):
+        """
+        Perform one incremental z update.
+        This is the "main" function of the algorithm.
+        """
+        raise NotImplementedError()
+
+    def compute_z_partial(self, i0):
+        """
+        Compute z on a slice of the signal X,
+        for online learning.
+
+        Parameters
+        ----------
+        i0 : int
+             Slice index.
+        """
+        raise NotImplementedError()
+
+    def get_cost(self):
+        """
+        Computes the cost of the current sparse representation (z_hat)
+
+        Returns
+        -------
+        cost: float
+        """
+        raise NotImplementedError()
+
+    def get_sufficient_statistics(self):
+        """
+        Computes sufficient statistics to update D.
+
+        Returns
+        -------
+        ztz, ztX : (float, float)
+            Sufficient statistics.
+        """
+        raise NotImplementedError()
+
+    def get_sufficient_statistics_partial(self):
+        """
+        Returns the partial sufficient statistics
+        that were computed during the last call to
+        compute_z_partial.
+
+        Returns
+        -------
+        ztz, ztX : (float, float)
+            Sufficient statistics for the slice that was
+            selected in the last call of ``compute_z_partial``
+        """
+        raise NotImplementedError()
+
+    def get_z_hat(self):
+        """
+        Returns a sparse encoding of the signal.
+
+        Returns
+        -------
+        z_hat
+            Sparse encoding of the signal X.
+        """
+        raise NotImplementedError()
+
+    def set_D(self, d):
+        """
+        Update the dictionary.
+
+        Parameters
+        ----------
+        d
```

Comment:
Maybe better to call this `D` ?
```suggestion
    def set_D(self, D):
        """
        Update the dictionary.

        Parameters
        ----------
        D : ndarray, shape (n_atoms, n_channels, n_time_atoms)
```